metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -421,17 +423,30 @@ class MetalsLanguageServer(
             didCompileTarget(report)
             compilers.didCompile(report)
           },
-          changes => {
-            quickConnectToBuildServer().onComplete {
-              case Failure(e) =>
-                scribe.warn("Error refreshing build", e)
-              case Success(_) =>
-                scribe.info("Refreshed build after change")
-            }
+          onBuildTargetDidCompile = { target =>
+            treeView.onBuildTargetDidCompile(target)
+            worksheetProvider.onBuildTargetDidCompile(target)
           },
-          () => treeView,
-          () => worksheetProvider,
-          () => ammonite
+          onBuildTargetDidChangeFunc = { params =>
+            val (ammoniteChanges, otherChanges) =
+              params.getChanges.asScala.partition(
+                _.getTarget.getUri.isAmmoniteScript
+              )
+            if (ammoniteChanges.nonEmpty)
+              ammonite.importBuild().onComplete {
```

Comment:
Could we move it to quickConnectToBuildServer (or a new method that does it?)

Reply:
Do you mean moving the whole `onBuildTargetDidChangeFunc` value, or just this Ammonite thing? For the latter, if the `onComplete` handlers stay here, this won't save any space.

Reply:
I would say the whole `onBuildTargetDidChangeFunc`

Reply:
Done.